adapters/repos/db/replication.go

```diff
@@ -20,95 +20,148 @@ import (
 
 	"github.com/go-openapi/strfmt"
 	"github.com/semi-technologies/weaviate/adapters/repos/db/vector/noop"
+	"github.com/semi-technologies/weaviate/entities/schema"
 	"github.com/semi-technologies/weaviate/entities/storobj"
 	hnswent "github.com/semi-technologies/weaviate/entities/vectorindex/hnsw"
 	"github.com/semi-technologies/weaviate/usecases/objects"
 	"github.com/semi-technologies/weaviate/usecases/replica"
 )
 
-func (ind *Index) IncomingFilePutter(ctx context.Context, shardName,
-	filePath string,
-) (io.WriteCloser, error) {
-	localShard, ok := ind.Shards[shardName]
-	if !ok {
-		return nil, fmt.Errorf("shard %q does not exist locally", shardName)
-	}
-
-	return localShard.filePutter(ctx, filePath)
+type ReplicatedIndexFactory interface {
+	GetReplicatedIndex(className schema.ClassName) Replicator
 }
 
-func (s *Shard) filePutter(ctx context.Context,
-	filePath string,
-) (io.WriteCloser, error) {
-	// TODO: validate file prefix to rule out that we're accidentally writing
-	// into another shard
-	finalPath := filepath.Join(s.index.Config.RootPath, filePath)
-	f, err := os.Create(finalPath)
-	if err != nil {
-		return nil, fmt.Errorf("open file %q for writing: %w", filePath, err)
-	}
+type Replicator interface {
+	ReplicateObject(ctx context.Context, shardName, requestID string,
+		object *storobj.Object) replica.SimpleResponse
+	ReplicateObjects(ctx context.Context, shardName, requestID string,
+		objects []*storobj.Object) replica.SimpleResponse
+	ReplicateUpdate(ctx context.Context, shard, requestID string,
+		doc *objects.MergeDocument) replica.SimpleResponse
+	ReplicateDeletion(ctx context.Context, shardName, requestID string,
+		uuid strfmt.UUID) replica.SimpleResponse
+	ReplicateDeletions(ctx context.Context, shardName, requestID string,
+		docIDs []uint64, dryRun bool) replica.SimpleResponse
+	ReplicateReferences(ctx context.Context, shard, requestID string,
+		refs []objects.BatchReference) replica.SimpleResponse
+	CommitReplication(ctx context.Context, shard,
+		requestID string) interface{}
+	AbortReplication(ctx context.Context, shardName,
+		requestID string) interface{}
+}
 
-	return f, nil
+type ReplicatedIndex struct {
+	repo ReplicatedIndexFactory
 }
 
-func (ind *Index) IncomingCreateShard(ctx context.Context,
-	shardName string,
-) error {
-	// TODO: locking???
-	if _, ok := ind.Shards[shardName]; ok {
-		return fmt.Errorf("shard %q exists already", shardName)
+func NewReplicatedIndex(repo ReplicatedIndexFactory) *ReplicatedIndex {
+	return &ReplicatedIndex{
+		repo: repo,
 	}
+}
 
-	// TODO: metrics
-	s, err := NewShard(ctx, nil, shardName, ind)
-	if err != nil {
-		return err
+func (rii *ReplicatedIndex) ReplicateObject(ctx context.Context, indexName,
+	shardName, requestID string, object *storobj.Object,
+) replica.SimpleResponse {
+	index := rii.repo.GetReplicatedIndex(schema.ClassName(indexName))
```

Comment:
Probably we can even get rid of ReplicatedIndex and GetReplicatedIndex in favour of DB.
I suggest instead to let DB implement the replication interface (see example below).

```
func (db *DB) ReplicateUpdate(ctx context.Context, indexName,
	shardName, requestID string, mergeDoc *objects.MergeDocument,
) replica.SimpleResponse {
	index := db.GetIndex(schema.ClassName(indexName))
	if index == nil {
		return replica.SimpleResponse{
			Errors: []string{fmt.Sprintf("local index %q not found", indexName)},
		}
	}

	return index.ReplicateUpdate(ctx, shardName, requestID, mergeDoc)
}


```